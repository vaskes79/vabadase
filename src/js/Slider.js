class Slider {
  constructor(
    slider = '.slider',
    btnL = '.slider__left-btn',
    btnR = '.slider__right-btn',
    slides = '.slider__slide',
    activeModifire = 'slider__slide--active'
  ) {
    this.slider = document.querySelector(slider);
    this.btnL = this.slider.querySelector(btnL);
    this.btnR = this.slider.querySelector(btnR);
    this.slides = this.slider.querySelectorAll(slides);
    this.slidesContainer = this.slides[0].parentNode;
    this.activeSlide = this.slider.querySelector(`.${activeModifire}`);
    this.activeModifire = activeModifire;
    this.angledeg = 0;
  }

  rotateContainer = angledeg => {
    this.slidesContainer.style.cssText = `
        -webkit-transform: rotateY(${angledeg}deg);
        -moz-transform: rotateY(${angledeg}deg);
        -o-transform: rotateY(${angledeg}deg);
        -ms-transform: rotateY(${angledeg}deg);
        transform: rotateY(${angledeg}deg);
    `;
  };

  handlerBtnL = () => {
    let currentSlide = this.activeSlide;
    let angledeg = this.angledeg;
    currentSlide.classList.toggle(this.activeModifire);

    if (this.slidesContainer.firstChild === this.activeSlide) {
      angledeg = -180 * this.slides.length + 180;
      currentSlide = this.slidesContainer.lastChild;
    } else {
      angledeg = angledeg + 180;
      currentSlide = this.activeSlide.previousSibling;
    }

    currentSlide.classList.toggle(this.activeModifire);
    this.rotateContainer(angledeg);
    this.angledeg = angledeg;
    this.activeSlide = currentSlide;
  };

  handlerBtnR = () => {
    let currentSlide = this.activeSlide;
    let angledeg = this.angledeg;
    currentSlide.classList.toggle(this.activeModifire);

    if (this.slidesContainer.lastChild === this.activeSlide) {
      angledeg = 0;
      currentSlide = this.slidesContainer.firstChild;
    } else {
      angledeg = angledeg - 180;
      currentSlide = this.activeSlide.nextSibling;
    }

    currentSlide.classList.toggle(this.activeModifire);
    this.rotateContainer(angledeg);
    this.angledeg = angledeg;
    this.activeSlide = currentSlide;
  };

  setHandlers() {
    this.btnL.addEventListener('click', this.handlerBtnL);
    this.btnR.addEventListener('click', this.handlerBtnR);
  }

  init() {
    this.setHandlers();
  }
}

export default Slider;
