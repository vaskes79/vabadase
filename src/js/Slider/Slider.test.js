import Slider from './Slider';
let slider = null;
let mock = null;
let clickHandlerMockL = null;
let clickHandlerMockR = null;

beforeEach(() => {
  slider = new Slider();
  clickHandlerMockL = jest.spyOn(slider, 'handlerBtnL');
  clickHandlerMockR = jest.spyOn(slider, 'handlerBtnR');
  slider.init();
  mock = jest.fn();
});

afterEach(() => {
  slider = null;
  mock = null;
  clickHandlerMockL = null;
  clickHandlerMockR = null;
});

describe(`Create instance and render`, () => {
  it(`expect that slider instance of Slider`, () => {
    expect(slider instanceof Slider).toBeTruthy();
  });

  it(`render components into DOM`, () => {
    expect(slider).toMatchSnapshot();
  });
});

describe(`props`, () => {
  describe(`default props`, () => {
    it(`this.slider exists has class .slider`, () => {
      let elem = slider.slider;
      expect(elem).not.toBeNil();
      expect(elem.classList.contains(`slider`)).toBeTrue();
    });

    it(`this.btnL exists has class .slider__left-btn`, () => {
      let elem = slider.btnL;
      expect(elem).not.toBeNil();
      expect(elem.classList.contains(`slider__left-btn`)).toBeTrue();
    });

    it(`this.btnR exists has class .slider__right-btn`, () => {
      let elem = slider.btnR;
      expect(elem).not.toBeNil();
      expect(elem.classList.contains(`slider__right-btn`)).toBeTrue();
    });

    it(`this.slides collection exists has class .slider__slide`, () => {
      let elem = slider.slides;
      expect(elem).not.toBeNil();

      [].forEach.call(elem, slide => {
        expect(slide.classList.contains('slider__slide')).toBeTrue();
      });
    });

    it(`this.slidesContainer exists has class .slider__slides-container`, () => {
      let elem = slider.slidesContainer;
      expect(elem).not.toBeNil();
      expect(elem.classList.contains(`slider__slides-container`)).toBeTrue();
    });

    it(`this.activeSlide exists has class .slider__slide--active`, () => {
      let elem = slider.activeSlide;
      expect(elem).not.toBeNil();
      expect(elem.classList.contains(`slider__slide--active`)).toBeTrue();
    });

    it('this.activeModifire exists has value slider__slide--active', () => {
      let elem = slider.activeModifire;
      expect(elem).not.toBeNil();
      expect(elem).toBeString();
      expect(elem).toBe('slider__slide--active');
    });
  });
});

describe('methods', () => {
  const listMethods = [
    'handlerBtnL',
    'handlerBtnR',
    'rotateContainer',
    'setHandlers',
    'init',
  ];

  listMethods.forEach(method => {
    describe(`${method}`, () => {
      it(`${method} exists`, () => {
        expect(slider[method]).toBeFunction();
      });

      it(`${method} toBeCalled`, () => {
        slider[method] = mock;
        slider[method]();
        expect(slider[method]).toBeCalled();
      });

      if (method === 'rotateContainer') {
        it(`set angle deg`, () => {
          let elem = slider.slidesContainer;
          let style = getComputedStyle(elem);
          let tr = style.transform;
          expect(tr).toBeEmpty();

          slider.rotateContainer(180);
          style = getComputedStyle(elem);
          tr = style.transform;
          expect(tr).toEqual('rotateY(180deg)');
        });
      }
    });
  });
});

describe(`events`, () => {
  it(`slider.btnL: click`, () => {
    let elem = slider.btnL;

    simulaitClick(elem);
    expect(clickHandlerMockL).toBeCalled();
    expect(clickHandlerMockL).toHaveBeenCalledTimes(1);
  });

  it(`slider.btnR: click`, () => {
    let elem = slider.btnR;

    simulaitClick(elem);
    expect(clickHandlerMockR).toBeCalled();
    expect(clickHandlerMockR).toHaveBeenCalledTimes(1);
  });
});
