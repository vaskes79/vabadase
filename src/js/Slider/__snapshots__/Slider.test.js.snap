// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create instance and render render components into DOM 1`] = `
Slider {
  "activeModifire": "slider__slide--active",
  "activeSlide": <figure
    class="slider__slide slider__slide--active"
  >
    <img
      allt="2 часа от петербурга"
      class="slider__slide-img"
      sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
      src="/static/img/c89553-slide-1-2000.jpg"
      srcset="/static/img/c89553-slide-1-2000.jpg 2000w"
    />
    <figcaption
      class="slider__slide-description"
    >
      <h3
        class="heading-3"
      >
        2 часа от петербурга
      </h3>
      <p
        class="paragraph"
      >
        Все лишь за 2 часа езды на машине можно добраться, до вашего дома в Европе. Где вас ждут Тихие ухоженные тропинки, сосновый лес и много чистого воздуха. 
      </p>
      <p
        class="paragraph"
      >
        Усть Нарва давно пользуется популярностью у писателей, художников, актеров и людей с творческим складом ума, здесь они любят отдохнуть от суеты и погрузится в гармонию сдешних мест.
      </p>
    </figcaption>
  </figure>,
  "angledeg": 0,
  "btnL": <a
    class="slider__left-btn"
    href="#previos"
  >
    <svg
      class="va-icon"
    >
       
      <use
        xlink:href="#va-icon-chevron-left"
      />
    </svg>
  </a>,
  "btnR": <a
    class="slider__right-btn"
    href="#next"
  >
    <svg
      class="va-icon"
    >
       
      <use
        xlink:href="#va-icon-chevron-right"
      />
    </svg>
  </a>,
  "handlerBtnL": [MockFunction],
  "handlerBtnR": [MockFunction],
  "rotateContainer": [Function],
  "slider": <div
    class="slider"
  >
    <a
      class="slider__left-btn"
      href="#previos"
    >
      <svg
        class="va-icon"
      >
         
        <use
          xlink:href="#va-icon-chevron-left"
        />
      </svg>
    </a>
    <div
      class="slider__slides-container"
    >
      <figure
        class="slider__slide slider__slide--active"
      >
        <img
          allt="2 часа от петербурга"
          class="slider__slide-img"
          sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
          src="/static/img/c89553-slide-1-2000.jpg"
          srcset="/static/img/c89553-slide-1-2000.jpg 2000w"
        />
        <figcaption
          class="slider__slide-description"
        >
          <h3
            class="heading-3"
          >
            2 часа от петербурга
          </h3>
          <p
            class="paragraph"
          >
            Все лишь за 2 часа езды на машине можно добраться, до вашего дома в Европе. Где вас ждут Тихие ухоженные тропинки, сосновый лес и много чистого воздуха. 
          </p>
          <p
            class="paragraph"
          >
            Усть Нарва давно пользуется популярностью у писателей, художников, актеров и людей с творческим складом ума, здесь они любят отдохнуть от суеты и погрузится в гармонию сдешних мест.
          </p>
        </figcaption>
      </figure>
      <figure
        class="slider__slide"
      >
        <img
          allt="3 минуты до моря"
          class="slider__slide-img"
          sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
          src="/static/img/c84cf5-slide-2-2000.jpg"
          srcset="/static/img/c84cf5-slide-2-2000.jpg 2000w"
        />
        <figcaption
          class="slider__slide-description"
        >
          <h3
            class="heading-3"
          >
            3 минуты до моря
          </h3>
          <p
            class="paragraph"
          >
            Дом VABADUSE расположен на побережъе нарвского залива, в трех минутах хотьбы от берега. Вы сможете очутится на море под теплым солнцем, буквально сразу как захотите!
          </p>
        </figcaption>
      </figure>
      <figure
        class="slider__slide"
      >
        <img
          allt="рядом все необходимое"
          class="slider__slide-img"
          sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
          src="/static/img/5fa1f7-slide-3-2000.jpg"
          srcset="/static/img/5fa1f7-slide-3-2000.jpg 2000w"
        />
        <figcaption
          class="slider__slide-description"
        >
          <h3
            class="heading-3"
          >
            рядом все необходимое
          </h3>
          <p
            class="paragraph"
          >
            Через дорогу от дома расположен супермаркет. В минуте хотьбы СПА санаторий, известный на всю эстонию своими лечебыными процедурами, «Нарва-Йыэсуу». В 2014 году построен современный СПА центр «Норус». Планируется так же строительство аквапарка.
          </p>
        </figcaption>
      </figure>
    </div>
    <a
      class="slider__right-btn"
      href="#next"
    >
      <svg
        class="va-icon"
      >
         
        <use
          xlink:href="#va-icon-chevron-right"
        />
      </svg>
    </a>
  </div>,
  "slides": NodeList [
    <figure
      class="slider__slide slider__slide--active"
    >
      <img
        allt="2 часа от петербурга"
        class="slider__slide-img"
        sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
        src="/static/img/c89553-slide-1-2000.jpg"
        srcset="/static/img/c89553-slide-1-2000.jpg 2000w"
      />
      <figcaption
        class="slider__slide-description"
      >
        <h3
          class="heading-3"
        >
          2 часа от петербурга
        </h3>
        <p
          class="paragraph"
        >
          Все лишь за 2 часа езды на машине можно добраться, до вашего дома в Европе. Где вас ждут Тихие ухоженные тропинки, сосновый лес и много чистого воздуха. 
        </p>
        <p
          class="paragraph"
        >
          Усть Нарва давно пользуется популярностью у писателей, художников, актеров и людей с творческим складом ума, здесь они любят отдохнуть от суеты и погрузится в гармонию сдешних мест.
        </p>
      </figcaption>
    </figure>,
    <figure
      class="slider__slide"
    >
      <img
        allt="3 минуты до моря"
        class="slider__slide-img"
        sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
        src="/static/img/c84cf5-slide-2-2000.jpg"
        srcset="/static/img/c84cf5-slide-2-2000.jpg 2000w"
      />
      <figcaption
        class="slider__slide-description"
      >
        <h3
          class="heading-3"
        >
          3 минуты до моря
        </h3>
        <p
          class="paragraph"
        >
          Дом VABADUSE расположен на побережъе нарвского залива, в трех минутах хотьбы от берега. Вы сможете очутится на море под теплым солнцем, буквально сразу как захотите!
        </p>
      </figcaption>
    </figure>,
    <figure
      class="slider__slide"
    >
      <img
        allt="рядом все необходимое"
        class="slider__slide-img"
        sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
        src="/static/img/5fa1f7-slide-3-2000.jpg"
        srcset="/static/img/5fa1f7-slide-3-2000.jpg 2000w"
      />
      <figcaption
        class="slider__slide-description"
      >
        <h3
          class="heading-3"
        >
          рядом все необходимое
        </h3>
        <p
          class="paragraph"
        >
          Через дорогу от дома расположен супермаркет. В минуте хотьбы СПА санаторий, известный на всю эстонию своими лечебыными процедурами, «Нарва-Йыэсуу». В 2014 году построен современный СПА центр «Норус». Планируется так же строительство аквапарка.
        </p>
      </figcaption>
    </figure>,
  ],
  "slidesContainer": <div
    class="slider__slides-container"
  >
    <figure
      class="slider__slide slider__slide--active"
    >
      <img
        allt="2 часа от петербурга"
        class="slider__slide-img"
        sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
        src="/static/img/c89553-slide-1-2000.jpg"
        srcset="/static/img/c89553-slide-1-2000.jpg 2000w"
      />
      <figcaption
        class="slider__slide-description"
      >
        <h3
          class="heading-3"
        >
          2 часа от петербурга
        </h3>
        <p
          class="paragraph"
        >
          Все лишь за 2 часа езды на машине можно добраться, до вашего дома в Европе. Где вас ждут Тихие ухоженные тропинки, сосновый лес и много чистого воздуха. 
        </p>
        <p
          class="paragraph"
        >
          Усть Нарва давно пользуется популярностью у писателей, художников, актеров и людей с творческим складом ума, здесь они любят отдохнуть от суеты и погрузится в гармонию сдешних мест.
        </p>
      </figcaption>
    </figure>
    <figure
      class="slider__slide"
    >
      <img
        allt="3 минуты до моря"
        class="slider__slide-img"
        sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
        src="/static/img/c84cf5-slide-2-2000.jpg"
        srcset="/static/img/c84cf5-slide-2-2000.jpg 2000w"
      />
      <figcaption
        class="slider__slide-description"
      >
        <h3
          class="heading-3"
        >
          3 минуты до моря
        </h3>
        <p
          class="paragraph"
        >
          Дом VABADUSE расположен на побережъе нарвского залива, в трех минутах хотьбы от берега. Вы сможете очутится на море под теплым солнцем, буквально сразу как захотите!
        </p>
      </figcaption>
    </figure>
    <figure
      class="slider__slide"
    >
      <img
        allt="рядом все необходимое"
        class="slider__slide-img"
        sizes="(max-width: 400px) 280px, (max-width: 600px) 440px, (max-width: 1200px) 800px, (max-width: 2000px) 1970px"
        src="/static/img/5fa1f7-slide-3-2000.jpg"
        srcset="/static/img/5fa1f7-slide-3-2000.jpg 2000w"
      />
      <figcaption
        class="slider__slide-description"
      >
        <h3
          class="heading-3"
        >
          рядом все необходимое
        </h3>
        <p
          class="paragraph"
        >
          Через дорогу от дома расположен супермаркет. В минуте хотьбы СПА санаторий, известный на всю эстонию своими лечебыными процедурами, «Нарва-Йыэсуу». В 2014 году построен современный СПА центр «Норус». Планируется так же строительство аквапарка.
        </p>
      </figcaption>
    </figure>
  </div>,
}
`;
